runtime-transaction/src/instruction_processor.rs

```diff
@@ -0,0 +1,62 @@
+use {
+    crate::program_id_flags::ProgramIdFlag,
+    agave_transaction_view::static_account_keys_frame::MAX_STATIC_ACCOUNTS_PER_PACKET as FILTER_SIZE,
+    core::borrow::Borrow, solana_pubkey::Pubkey, solana_sdk::transaction::TransactionError,
+    solana_svm_transaction::instruction::SVMInstruction,
+};
+
+pub trait InstructionProcessor<T>: Default {
+    type OUTPUT;
+
+    /// Process instructions individually given a `flag` that is derived from
+    /// the program id.
+    fn process_instruction(
+        &mut self,
+        flag: &T,
+        instruction_index: usize,
+        instruction: &SVMInstruction,
+    ) -> Result<(), TransactionError>;
+
+    /// Finalize the processor and return the output.
+    fn finalize(self) -> Self::OUTPUT;
+}
+
+pub struct CachedInstructionProcessor<T, P: InstructionProcessor<T>> {
+    flags: [Option<T>; FILTER_SIZE as usize],
+    processor: P,
+}
+
+impl<T, P> Default for CachedInstructionProcessor<T, P>
+where
+    P: InstructionProcessor<T> + Default,
+{
+    #[inline]
+    fn default() -> Self {
+        Self {
+            flags: core::array::from_fn(|_| None),
+            processor: P::default(),
+        }
+    }
+}
+
+impl<T, P> CachedInstructionProcessor<T, P>
+where
+    T: for<'k> From<&'k ProgramIdFlag>,
+    P: InstructionProcessor<T>,
+{
+    #[inline]
+    pub fn process_instructions<'b>(
```

Comment:
clever but a bit over-engineering to me. The function fits nicely with Visitor design pattern, I personally prefer Visitor if possible, for its readability and maintainability